torch/csrc/autograd/profiler_utils.cpp

```diff
@@ -0,0 +1,201 @@
+#include <torch/csrc/autograd/profiler_utils.h>
+
+namespace torch { namespace autograd { namespace profiler {
+
+const size_t CONV2D_STRIDE = 3;
+const size_t CONV2D_PADDING = 4;
+const size_t CONV2D_DILATION = 5;
+const size_t CONV2D_GROUPS = 6;
+
+static bool validateInput(const at::StringView &op_name, size_t min_size,
+                       const std::vector<c10::IValue>& inputs,
+                       const std::vector<int>& should_be_tensor) {
+  std::stringstream ss;
+  if (inputs.size() < min_size) {
+      ss << "Failed to save extra arguments for flops compuation of op "
+         << op_name.str()
+         << ", min size: " << min_size
+         << ", actual size: " << inputs.size();
+      TORCH_WARN(ss.str());
+      return false;
+  }
+  for (auto index : should_be_tensor) {
+    if (!inputs[index].isTensor()) {
+      ss << "Failed to save extra arguments for flops compuation of op "
+         << op_name.str()
+         << ", input[" << index
+         << "] must be a tensor.";
+      TORCH_WARN(ss.str());
+      return false;
+    }
+  }
+  return true;
+}
+
+std::unordered_map<std::string, c10::IValue> saveExtraArgs(const at::RecordFunction& fn) {
```

Comment:
can it be simplified? e.g. remember shapes for all tensor inputs, record all int args and have a whitelist of operators for which your record all strings args. See D24472855 which tries to do something similar for logging purposes

You don't really need to validate the args here - you can assume that pytorch is self-correct. If anything, you can just have one catch all try/catch

Reply:
once we just copy these ints, we'd still need to figure which one is which? so most of the logic seems to stay anyway?

Reply:
i think the main difference between here and D24472855 is that D24472855 just outputs these ints to scuba but here we use them non-trivially to compute flops, hence extra code

Reply:
My point is that you're duplicating the op schema in one more place (by hard-coding the argument names and order), it's not maintainable if there going to be many flops functions.

You can make this function generic by just iterating over inputs, storing shapes for tensors and values for non-tensors. Then the saving part is generic and you'd only need to do the parsing part in one place - in analysis.

And you don't need to do fancy error checking here, you can rely that the inputs match the schema (because otherwise the call would fail in unboxing logic trying to call the kernel anyway)

Reply:
so, you're saying that saveExtraArgs should be as simple as copying scalars, if I understand correctly.
that's fine to me, my point was saying that this parsing we still need to do, then in computeFlops, right? so some form of schema verification duplication still going to be there?

Reply:
I like the idea of a tight integration of operator implementation and flops formula, we need to follow up on the design and add this to op registration (in follow up pr)